pkg/pillar/dpcreconciler/linux.go

```diff
@@ -2334,3 +2384,259 @@ func defaultKubernetesIptablesRules() []iptables.Rule {
+				} else {
+					// For LoadBalancer without IP, add to generic port list
+					udpPortsMap[strconv.Itoa(int(svc.Port))] = true
+				}
+			}
+		}
+	}
+
+	// Create IP-specific rules for UDP LoadBalancer services
+	ruleIndex = 0
+	for ip, ports := range udpLoadBalancerIPRules {
+		var udpPorts []string
+		for port := range ports {
+			udpPorts = append(udpPorts, port)
+		}
+
+		// If there are ports for this IP
+		if len(udpPorts) > 0 {
+			markUDPSvcPortIP := iptables.Rule{
+				RuleLabel:   fmt.Sprintf("KubeSvcPortUDP_IP_%d", ruleIndex),
+				MatchOpts:   []string{"-p", "udp", "-d", ip, "--match", "multiport", "--dports", strings.Join(udpPorts, ",")},
+				Target:      "CONNMARK",
+				TargetOpts:  []string{"--set-mark", controlProtoMark("in_udp_svc_port")},
+				Description: fmt.Sprintf("Mark Kubernetes UDP service port traffic for LoadBalancer IP %s", ip),
+			}
+			rules = append(rules, markUDPSvcPortIP)
+			ruleIndex++
+		}
+	}
+
+	// Create generic rule for UDP NodePorts and LoadBalancer services without IPs
+	if len(udpPortsMap) > 0 {
+		var udpPorts []string
+		for port := range udpPortsMap {
+			udpPorts = append(udpPorts, port)
+		}
+
+		markUDPSvcPort := iptables.Rule{
+			RuleLabel:   "KubeSvcPortUDP",
+			MatchOpts:   []string{"-p", "udp", "--match", "multiport", "--dports", strings.Join(udpPorts, ",")},
+			Target:      "CONNMARK",
+			TargetOpts:  []string{"--set-mark", controlProtoMark("in_udp_svc_port")},
+			Description: "Mark Kubernetes UDP service port traffic",
+		}
+		rules = append(rules, markUDPSvcPort)
+	}
+
+	// 3. Create rules for HTTP and HTTPS ingresses
+	// Track which protocols are used and collect unique ingress IPs
+	httpIngressIPs := make(map[string]bool)
+	httpsIngressIPs := make(map[string]bool)
+	var hasHTTPNoIP bool
+	var hasHTTPSNoIP bool
+
+	// Collect information about ingresses, unique IPs and protocols
+	// Two cases, with IP and protocol or the protocol only for non-loadBalancer services
+	for _, ing := range kubeUserServices.UserIngress {
+		// Only process ingress IPs for LoadBalancer services
+		// For NodePort types, the addresses will be cluster-prefixes which we use 10.244.244.0
+		// private addresses, and user may not have access to them.
+		if ing.ServiceType == corev1.ServiceTypeLoadBalancer {
+			// Process HTTP protocol
+			if ing.Protocol == "http" || ing.Protocol == "https" {
+				// Process IPs for LoadBalancer
+				for _, ip := range ing.IngressIP {
+					if ip == "" {
+						continue
+					}
+					if ing.Protocol == "http" {
+						httpIngressIPs[ip] = true
+					} else { // https
+						httpsIngressIPs[ip] = true
+					}
+				}
+			}
+			continue
+		}
+
+		// For NodePort services, we only track the protocol number
+		if ing.Protocol == "http" {
+			// For NodePort HTTP, we don't track IPs, just mark the port
+			hasHTTPNoIP = true
+		} else if ing.Protocol == "https" {
+			// For NodePort HTTPS, we don't track IPs, just mark the port
+			hasHTTPSNoIP = true
+		}
```

Comment:
I think you can rebalance this a bit to have less deep nesting:

```suggestion
		// Process HTTP protocol
		if ing.ServiceType == corev1.ServiceTypeLoadBalancer && (ing.Protocol == "http" || ing.Protocol == "https") {
			// Process IPs for LoadBalancer
			for _, ip := range ing.IngressIP {
				if ip == "" {
					continue
				}
				if ing.Protocol == "http" {
					httpIngressIPs[ip] = true
				} else { // https
					httpsIngressIPs[ip] = true
				}
			}
		}
		// For NodePort services, we only track the protocol number			
		} else if ing.Protocol == "http" {
			// For NodePort HTTP, we don't track IPs, just mark the port
			hasHTTPNoIP = true
		} else if ing.Protocol == "https" {
			// For NodePort HTTPS, we don't track IPs, just mark the port
			hasHTTPSNoIP = true
		}
```

Reply:
(I hope I did not add any mistakes - but I hope the tests would find those)

Reply:
sure.